drivers/adc/adc_mcp356x.c

```diff
@@ -0,0 +1,836 @@
+ * interval_us delay.  Channel 16 does not run in continuous mode and can not sample at as high
+ * of a rate.  This is not just because of extra SPI commands and more overhead.  The manner in
+ * which a ΔΣ ADC creates the samples is different between continuous conversion and a sequence
+ * of one-shot conversions.
+ *
+ * Setting calibrate in the adc_sequence will enable the offset cancellatin algorithm, see
+ * §5.1.3, which causes the ADC to sample twice, with the mux set in either direction, to
+ * attempt to cancel out offset error.  This doubles the time each sample takes.
+ */
+
+/* Missing features.
+ *
+ * Configure bias current (in device tree?).
+ * Configure internal clock output to MCLK pin (dts).
+ * Standby vs shutdown when not sampling (dts?).
+ * Tweak IRQ timeout to just to expected sampling rate.
+ * Support the CRC data checking mode.
+ * Detect POR or unexpected config change via status bits when reading data.
+ * Timestamp the DRDY interrupt, to get more accurate data timestamps.
+ */
+
+#include <drivers/adc.h>
+#include <drivers/gpio.h>
+#include <drivers/spi.h>
+#include <kernel.h>
+#include <logging/log.h>
+#include <sys/byteorder.h>
+#include <sys/util.h>
+#include <zephyr.h>
+
+#include "adc_mcp356x.h"
+
+LOG_MODULE_REGISTER(adc_mcp356x, CONFIG_ADC_LOG_LEVEL);
+
+/* Verbose debug logging.  Individual register access. IRQ timing. */
+#define VERBOSE_LOG	0
+
+#if VERBOSE_LOG
+#define LOG_VDBG(...)	LOG_DBG(__VA_ARGS__)
+#else
+#define LOG_VDBG(...)	(void)(0)
+#endif
+
+#if !IS_ENABLED(CONFIG_ADC_MCP356X_USE_INTERNAL_TIMER)
+#define ADC_CONTEXT_USES_KERNEL_TIMER
+#endif
+#include "adc_context.h"
+
+struct mcp356x_config {
+	struct spi_dt_spec bus;
+	struct gpio_dt_spec irq;
+	unsigned int mclk_freq;	/* External or internal clock rate */
+	uint8_t channels;	/* Number of channels */
+	uint8_t addr;		/* Device address selection bits */
+	uint8_t pre;		/* Prescaler */
+	uint8_t boost;		/* Boost setting (register value) */
+	bool ext_clock;		/* Using external or internal clock */
+	bool push_pull;		/* Drive mode of interrupt pin */
+};
+
+struct mcp356x_data {
+	struct adc_context ctx;		/* Must have this, ADC_CONTEXT_INIT macros use it */
+	struct gpio_callback drdy_cb;	/* For data ready IRQ */
+	struct k_thread thread;		/* Acquisition thread */
+	struct k_sem acq_sem;		/* Signal acq thread for next sample */
+	struct k_sem drdy_sem;		/* Signal data ready IRQ */
+
+	uint8_t *write_ptr;		/* Current address to send data */
+
+	uint16_t ch_mask;		/* Allow channels, based on chip type */
+	uint8_t delay;			/* Inter-channel delay (register code >> 16) */
+	uint8_t resolution;		/* Current resolution (in bits) */
+	uint8_t osr;			/* Ovesampling (register code, not shifted) */
+	uint8_t gain;			/* Gain (register code) */
+	bool az_mux;			/* Auto-zero mux */
+
+	K_KERNEL_STACK_MEMBER(stack,
+			CONFIG_ADC_MCP356X_ACQUISITION_THREAD_STACK_SIZE);
+};
+
+/* There are functions for unaligned be24 and le24, but none for CPU endian! */
```

Comment:
Could we add these instead of working around them not being present?

Reply:
Could *I* add them?  My experience with Zephyr so far is that this is unlikely.  Maybe someone who can put it on the agenda at Nordic's sprint planning would have better luck?